From e229a55d8c9763526aa2fa47e78b4558ef987dbf Mon Sep 17 00:00:00 2001
From: Patrice Chotard <patrice.chotard@foss.st.com>
Date: Tue, 17 Aug 2021 15:54:17 +0200
Subject: [PATCH 048/120] memory: Add STM32 Octo Memory Interface driver
 support

On SoC STM32MP2, a new OCTOSPI block is available.
This block supports both OSPI flashes and Hyperflashes.
This Octo Memory Interface driver (OMI) is first probed, and
depending on sub node compatible, it probes either the
STM32 OSPI driver or the STM32 Hyperbus sub-driver.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git 641cf2c7352b4e7e7d651a5e8e0195e5a5f360b8 memory: Add STM32 Octo Memory Interface driver support]
Signed-off-by: Patrice Chotard <patrice.chotard@foss.st.com>
Change-Id: Ied4c46d4455061e86196162ea8c2046d9adbb33f
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/230297
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Christophe KERELLO <christophe.kerello@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/308512
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Christophe KERELLO <christophe.kerello@st.com>
---
 drivers/memory/Kconfig     |   8 ++
 drivers/memory/Makefile    |   1 +
 drivers/memory/stm32-omi.c | 227 +++++++++++++++++++++++++++++++++++++
 include/stm32_omi.h        | 140 +++++++++++++++++++++++
 4 files changed, 376 insertions(+)
 create mode 100644 drivers/memory/stm32-omi.c
 create mode 100644 include/stm32_omi.h

diff --git a/drivers/memory/Kconfig b/drivers/memory/Kconfig
index 7c40f176987..bafd3534c2a 100644
--- a/drivers/memory/Kconfig
+++ b/drivers/memory/Kconfig
@@ -37,6 +37,14 @@ config STM32_FMC2_EBI
 	  devices (like SRAM, ethernet adapters, FPGAs, LCD displays, ...) on
 	  SOCs containing the FMC2 External Bus Interface.
 
+config STM32_OMI
+	bool "Support for Octo Memory Interface on STM32MP SoCs"
+	depends on ARCH_STM32MP
+	help
+	  Select this option to enable the STM32 Octo Memory Interface controller
+	  (OMI) which provides either SPI or HyperBus support. This allows to support
+	  sNOR, sNAND or HyperFlash devices.
+
 config TI_AEMIF
 	tristate "Texas Instruments AEMIF driver"
 	depends on ARCH_KEYSTONE || ARCH_DAVINCI
diff --git a/drivers/memory/Makefile b/drivers/memory/Makefile
index fdc83e4e1c8..aea57ce758a 100644
--- a/drivers/memory/Makefile
+++ b/drivers/memory/Makefile
@@ -2,6 +2,7 @@
 obj-$(CONFIG_MEMORY) += memory-uclass.o
 obj-$(CONFIG_SANDBOX_MEMORY) += memory-sandbox.o
 obj-$(CONFIG_STM32_FMC2_EBI) += stm32-fmc2-ebi.o
+obj-$(CONFIG_STM32_OMI) += stm32-omi.o
 obj-$(CONFIG_ATMEL_EBI) += atmel_ebi.o
 obj-$(CONFIG_TI_AEMIF) += ti-aemif.o ti-aemif-cs.o
 obj-$(CONFIG_TI_GPMC) += ti-gpmc.o
diff --git a/drivers/memory/stm32-omi.c b/drivers/memory/stm32-omi.c
new file mode 100644
index 00000000000..a8167089bef
--- /dev/null
+++ b/drivers/memory/stm32-omi.c
@@ -0,0 +1,227 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR BSD-3-Clause
+/*
+ * Copyright (C) 2021, STMicroelectronics - All Rights Reserved
+ */
+
+#include <dm.h>
+#include <log.h>
+#include <misc.h>
+#include <stm32_omi.h>
+#include <watchdog.h>
+#include <dm/of_access.h>
+#include <dm/device-internal.h>
+#include <dm/device_compat.h>
+#include <dm/lists.h>
+#include <dm/read.h>
+#include <linux/ioport.h>
+
+static void stm32_omi_read_fifo(u8 *val, phys_addr_t addr)
+{
+	*val = readb(addr);
+	schedule();
+}
+
+static void stm32_omi_write_fifo(u8 *val, phys_addr_t addr)
+{
+	writeb(*val, addr);
+}
+
+int stm32_omi_tx_poll(struct stm32_omi_plat *omi, u8 *buf, u32 len, bool read)
+{
+	phys_addr_t regs_base = omi->regs_base;
+	void (*fifo)(u8 *val, phys_addr_t addr);
+	u32 sr;
+	int ret;
+
+	if (read)
+		fifo = stm32_omi_read_fifo;
+	else
+		fifo = stm32_omi_write_fifo;
+
+	while (len--) {
+		ret = readl_poll_timeout(regs_base + OSPI_SR, sr,
+					 sr & OSPI_SR_FTF,
+					 OSPI_FIFO_TIMEOUT_US);
+		if (ret) {
+			dev_err(omi->dev, "fifo timeout (len:%d stat:%#x)\n",
+				len, sr);
+			return ret;
+		}
+
+		fifo(buf++, regs_base + OSPI_DR);
+	}
+
+	return 0;
+}
+
+int stm32_omi_wait_for_not_busy(struct stm32_omi_plat *omi)
+{
+	phys_addr_t regs_base = omi->regs_base;
+	u32 sr;
+	int ret;
+
+	ret = readl_poll_timeout(regs_base + OSPI_SR, sr, !(sr & OSPI_SR_BUSY),
+				 OSPI_BUSY_TIMEOUT_US);
+	if (ret)
+		dev_err(omi->dev, "busy timeout (stat:%#x)\n", sr);
+
+	return ret;
+}
+
+int stm32_omi_wait_cmd(struct stm32_omi_plat *omi)
+{
+	phys_addr_t regs_base = omi->regs_base;
+	u32 sr;
+	int ret = 0;
+
+	ret = readl_poll_timeout(regs_base + OSPI_SR, sr,
+				 sr & OSPI_SR_TCF,
+				 OSPI_CMD_TIMEOUT_US);
+	if (ret) {
+		dev_err(omi->dev, "cmd timeout (stat:%#x)\n", sr);
+	} else if (readl(regs_base + OSPI_SR) & OSPI_SR_TEF) {
+		dev_err(omi->dev, "transfer error (stat:%#x)\n", sr);
+		ret = -EIO;
+	}
+
+	/* clear flags */
+	writel(OSPI_FCR_CTCF | OSPI_FCR_CTEF, regs_base + OSPI_FCR);
+
+	if (!ret)
+		ret = stm32_omi_wait_for_not_busy(omi);
+
+	return ret;
+}
+
+static int stm32_omi_bind(struct udevice *dev)
+{
+	struct stm32_omi_plat *plat = dev_get_plat(dev);
+	struct driver *drv;
+	const char *name;
+	ofnode flash_node;
+	u8 hyperflash_count = 0;
+	u8 spi_flash_count = 0;
+	u8 child_count = 0;
+
+	/*
+	 * Flash subnodes sanity check:
+	 *        2 spi-nand/spi-nor flashes			=> supported
+	 *        1 HyperFlash					=> supported
+	 *	  All other flash node configuration		=> not supported
+	 */
+	dev_for_each_subnode(flash_node, dev) {
+		if (ofnode_device_is_compatible(flash_node, "cfi-flash"))
+			hyperflash_count++;
+
+		if (ofnode_device_is_compatible(flash_node, "jedec,spi-nor") ||
+		    ofnode_device_is_compatible(flash_node, "spi-nand"))
+			spi_flash_count++;
+
+		child_count++;
+	}
+
+	if (!child_count) {
+		dev_err(dev, "Missing flash node\n");
+		return -ENODEV;
+	}
+
+	if ((!hyperflash_count && !spi_flash_count) ||
+	    child_count != (hyperflash_count + spi_flash_count)) {
+		dev_warn(dev, "Unknown flash type\n");
+		return -ENODEV;
+	}
+
+	if ((hyperflash_count && spi_flash_count) ||
+	    hyperflash_count > 1) {
+		dev_err(dev, "Flash node configuration not supported\n");
+		return -EINVAL;
+	}
+
+	if (spi_flash_count)
+		name = "stm32_ospi";
+	else
+		name = "stm32_hyperbus";
+
+	drv = lists_driver_lookup_name(name);
+	if (!drv) {
+		dev_err(dev, "Cannot find driver '%s'\n", name);
+		return -ENOENT;
+	}
+
+	return device_bind(dev, drv, dev_read_name(dev), NULL, dev_ofnode(dev), NULL);
+}
+
+static int stm32_omi_of_to_plat(struct udevice *dev)
+{
+	struct stm32_omi_plat *plat = dev_get_plat(dev);
+	struct resource res;
+	struct ofnode_phandle_args args;
+	const fdt32_t *reg;
+	int ret, len;
+
+	reg = dev_read_prop(dev, "reg", &len);
+	if (!reg) {
+		dev_err(dev, "Can't get regs base address\n");
+		return -ENOENT;
+	}
+
+	plat->regs_base = (phys_addr_t)dev_translate_address(dev, reg);
+
+	/* optional */
+	ret = dev_read_phandle_with_args(dev, "memory-region", NULL, 0, 0, &args);
+	if (!ret) {
+		ret = ofnode_read_resource(args.node, 0, &res);
+		if (ret) {
+			dev_err(dev, "Can't get mmap base address(%d)\n", ret);
+			return ret;
+		}
+
+		plat->mm_base = res.start;
+		plat->mm_size = resource_size(&res);
+
+		if (plat->mm_size > OSPI_MAX_MMAP_SZ) {
+			dev_err(dev, "Incorrect memory-map size: %lld Bytes\n", plat->mm_size);
+			return -EINVAL;
+		}
+
+		dev_dbg(dev, "%s: regs_base=<0x%llx> mm_base=<0x%llx> mm_size=<0x%x>\n",
+			__func__, plat->regs_base, plat->mm_base, (u32)plat->mm_size);
+	} else {
+		plat->mm_base = 0;
+		plat->mm_size = 0;
+		dev_info(dev, "memory-region property not found (%d)\n", ret);
+	}
+
+	ret = clk_get_by_index(dev, 0, &plat->clk);
+	if (ret < 0) {
+		dev_err(dev, "Failed to get clock\n");
+		return ret;
+	}
+
+	ret = reset_get_bulk(dev, &plat->rst_ctl);
+	if (ret && ret != -ENOENT) {
+		dev_err(dev, "Failed to get reset\n");
+		return ret;
+	}
+
+	plat->clock_rate = clk_get_rate(&plat->clk);
+	if (!plat->clock_rate) {
+		return -EINVAL;
+	}
+
+	return 0;
+};
+
+static const struct udevice_id stm32_omi_ids[] = {
+	{.compatible = "st,stm32mp25-omi" },
+	{ }
+};
+
+U_BOOT_DRIVER(stm32_omi) = {
+	.name		= "stm32-omi",
+	.id		= UCLASS_NOP,
+	.of_match	= stm32_omi_ids,
+	.of_to_plat	= stm32_omi_of_to_plat,
+	.plat_auto	= sizeof(struct stm32_omi_plat),
+	.bind		= stm32_omi_bind,
+};
diff --git a/include/stm32_omi.h b/include/stm32_omi.h
new file mode 100644
index 00000000000..907f453f2a2
--- /dev/null
+++ b/include/stm32_omi.h
@@ -0,0 +1,140 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later OR BSD-3-Clause */
+/*
+ * Copyright (C) 2021, STMicroelectronics - All Rights Reserved
+ */
+
+#include <clk.h>
+#include <reset.h>
+#include <linux/iopoll.h>
+#include <linux/ioport.h>
+
+/*
+ * OCTOSPI control register
+ */
+#define OSPI_CR			0x00
+#define OSPI_CR_EN		BIT(0)
+#define OSPI_CR_ABORT		BIT(1)
+#define OSPI_CR_TCEN		BIT(3)
+#define OSPI_CR_FSEL		BIT(7)
+#define OSPI_CR_FTHRES_MASK	GENMASK(12,8)
+#define OSPI_CR_CSSEL		BIT(24)
+#define OSPI_CR_FMODE_SHIFT	28
+#define OSPI_CR_FMODE_MASK	GENMASK(29, 28)
+#define OSPI_CR_FMODE_IND_WRITE	0
+#define OSPI_CR_FMODE_IND_READ	1
+#define OSPI_CR_FMODE_MMAP	3
+/*
+ * OCTOSPI device configuration register
+ */
+#define OSPI_DCR1		0x08
+#define OSPI_DCR1_CKMODE	BIT(0)
+#define OSPI_DCR1_DLYBYP	BIT(3)
+#define OSPI_DCR1_CSHT_SHIFT	8
+#define OSPI_DCR1_CSHT_MASK	GENMASK(13, 8)
+#define OSPI_DCR1_DEVSIZE_MASK	GENMASK(20,16)
+#define OSPI_DCR1_MTYP_MASK	GENMASK(26, 24)
+#define OSPI_DCR1_MTYP_HP_MEMMODE	4
+/*
+ * OCTOSPI device configuration register 2
+ */
+#define OSPI_DCR2		0x0c
+#define OSPI_DCR2_PRESC_SHIFT	0
+#define OSPI_DCR2_PRESC_MASK	GENMASK(7, 0)
+/*
+ * OCTOSPI status register
+ */
+#define OSPI_SR			0x20
+#define OSPI_SR_TEF		BIT(0)
+#define OSPI_SR_TCF		BIT(1)
+#define OSPI_SR_FTF		BIT(2)
+#define OSPI_SR_BUSY		BIT(5)
+/*
+ * OCTOSPI flag clear register
+ */
+#define OSPI_FCR		0x24
+#define OSPI_FCR_CTEF		BIT(0)
+#define OSPI_FCR_CTCF		BIT(1)
+/*
+ * OCTOSPI data length register
+ */
+#define OSPI_DLR		0x40
+/*
+ * OCTOSPI address register
+ */
+#define OSPI_AR			0x48
+/*
+ * OCTOSPI data configuration register
+ */
+#define OSPI_DR			0x50
+/*
+ * OCTOSPI communication configuration register
+ */
+#define OSPI_CCR		0x100
+#define OSPI_CCR_IMODE_SHIFT	0
+#define OSPI_CCR_IMODE_MASK	GENMASK(2, 0)
+#define OSPI_CCR_ADMODE_SHIFT	8
+#define OSPI_CCR_ADMODE_MASK	GENMASK(10, 8)
+#define OSPI_CCR_ADMODE_8LINES	4
+#define OSPI_CCR_ADDTR		BIT(11)
+#define OSPI_CCR_ADSIZE_SHIFT	12
+#define OSPI_CCR_ADSIZE_MASK	GENMASK(13,12)
+#define OSPI_CCR_ADSIZE_32BITS	3
+#define OSPI_CCR_DMODE_SHIFT	24
+#define OSPI_CCR_DMODE_MASK	GENMASK(26, 24)
+#define OSPI_CCR_DMODE_8LINES	4
+#define OSPI_CCR_IND_WRITE	0
+#define OSPI_CCR_IND_READ	1
+#define OSPI_CCR_MEM_MAP	3
+#define OSPI_CCR_DDTR		BIT(27)
+#define OSPI_CCR_DQSE		BIT(29)
+/*
+ * OCTOSPI timing configuration register
+ */
+#define OSPI_TCR		0x108
+#define OSPI_TCR_DCYC_SHIFT	0x0
+#define OSPI_TCR_DCYC_MASK	GENMASK(4, 0)
+#define OSPI_TCR_DHQC		BIT(28)
+#define OSPI_TCR_SSHIFT		BIT(30)
+/*
+ * OCTOSPI instruction register
+ */
+#define OSPI_IR			0x110
+/*
+ * OCTOSPI low power timeout register
+ */
+#define OSPI_LPTR		0x130
+#define OSPI_LPTR_TIMEOUT_MASK	GENMASK(15, 0)
+
+/*
+ * OCTOSPI write communication configuration register
+ */
+#define OSPI_WCCR		0x180
+/*
+ * HyperBus latency configuration register
+ */
+#define OSPI_HLCR		0x200
+#define OSPI_HLCR_WZL		BIT(1)
+#define OSPI_HLCR_TACC_MASK	GENMASK(15,8)
+#define OSPI_HLCR_TRWR_MASK	GENMASK(23,16)
+
+#define OSPI_MAX_MMAP_SZ	SZ_256M
+#define OSPI_MAX_CHIP		2
+
+#define OSPI_ABT_TIMEOUT_US	100000
+#define OSPI_BUSY_TIMEOUT_US	100000
+#define OSPI_CMD_TIMEOUT_US	1000000
+#define OSPI_FIFO_TIMEOUT_US	30000
+
+struct stm32_omi_plat {
+	struct udevice *dev;
+	phys_addr_t regs_base;		/* register base address */
+	phys_addr_t mm_base;		/* memory map base address */
+	resource_size_t mm_size;
+	struct clk clk;
+	struct reset_ctl_bulk rst_ctl;
+	ulong clock_rate;
+};
+
+int stm32_omi_tx_poll(struct stm32_omi_plat *omi, u8 *buf, u32 len, bool read);
+int stm32_omi_wait_cmd(struct stm32_omi_plat *omi);
+int stm32_omi_wait_for_not_busy(struct stm32_omi_plat *omi);
-- 
2.47.2

